gradio/routes.py

```diff
@@ -628,17 +606,49 @@ async def join_queue(
                 app._asyncio_tasks.append(task)
             else:
                 rank = blocks._queue.push(event)
-
                 if rank is None:
-                    await blocks._queue.send_message(event, {"msg": "queue_full"})
-                    await event.disconnect()
+                    blocks._queue.send_message(event, "queue_full", final=True)
                     return
                 estimation = blocks._queue.get_estimation()
                 await blocks._queue.send_estimation(event, estimation, rank)
-            while True:
-                await asyncio.sleep(1)
-                if websocket.application_state == WebSocketState.DISCONNECTED:
-                    return
+
+            async def sse_stream(request: fastapi.Request):
+                while True:
+                    if await request.is_disconnected():
+                        await blocks._queue.clean_event(event)
+                    if not event.alive:
+                        return
+
+                    heartbeat_rate = 15
+                    check_rate = 0.05
+                    last_heartbeat = time.time()
+                    message = None
+                    try:
+                        message = event.message_queue.get_nowait()
+                        if message is None:  # end of stream marker
+                            return
+                    except EmptyQueue:
+                        await asyncio.sleep(check_rate)
+                        if time.time() - last_heartbeat > heartbeat_rate:
+                            message = {"msg": "heartbeat"}
+                            last_heartbeat = time.time()
+
+                    if message:
+                        yield f"data: {json.dumps(message)}\n\n"
+
+            return StreamingResponse(
+                sse_stream(request),
+                media_type="text/event-stream",
+            )
+
+        @app.post("/queue/data", dependencies=[Depends(login_check)])
+        async def queue_data(
+            body: PredictBody,
+            request: fastapi.Request,
+            username: str = Depends(get_current_user),
+        ):
+            blocks = app.get_blocks()
+            blocks._queue.attach_data(body)
```

Comment:
This request always succeeds even if we don't attach the data. I think we should return an appropriate status code and let the clients handle it.

Reply:
done